packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/BsonEncoder.kt

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2022 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@file:Suppress("invisible_reference", "invisible_member")
+
+package io.realm.kotlin.mongodb.internal
+
+import io.realm.kotlin.internal.ObjectIdImpl
+import io.realm.kotlin.internal.ObjectIdSerializer
+import io.realm.kotlin.internal.RealmInstantSerializer
+import io.realm.kotlin.internal.RealmUUIDSerializer
+import io.realm.kotlin.internal.toMillis
+import io.realm.kotlin.internal.toRealmInstant
+import io.realm.kotlin.types.MutableRealmInt
+import io.realm.kotlin.types.ObjectId
+import io.realm.kotlin.types.RealmInstant
+import io.realm.kotlin.types.RealmUUID
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.modules.plus
+import kotlinx.serialization.serializer
+import org.mongodb.kbson.BsonArray
+import org.mongodb.kbson.BsonBinary
+import org.mongodb.kbson.BsonBinarySubType
+import org.mongodb.kbson.BsonBoolean
+import org.mongodb.kbson.BsonDBPointer
+import org.mongodb.kbson.BsonDateTime
+import org.mongodb.kbson.BsonDecimal128
+import org.mongodb.kbson.BsonDocument
+import org.mongodb.kbson.BsonDouble
+import org.mongodb.kbson.BsonElement
+import org.mongodb.kbson.BsonInt32
+import org.mongodb.kbson.BsonInt64
+import org.mongodb.kbson.BsonInvalidOperationException
+import org.mongodb.kbson.BsonJavaScript
+import org.mongodb.kbson.BsonJavaScriptWithScope
+import org.mongodb.kbson.BsonMaxKey
+import org.mongodb.kbson.BsonMinKey
+import org.mongodb.kbson.BsonNull
+import org.mongodb.kbson.BsonNumber
+import org.mongodb.kbson.BsonObjectId
+import org.mongodb.kbson.BsonRegularExpression
+import org.mongodb.kbson.BsonString
+import org.mongodb.kbson.BsonSymbol
+import org.mongodb.kbson.BsonTimestamp
+import org.mongodb.kbson.BsonType
+import org.mongodb.kbson.BsonUndefined
+import org.mongodb.kbson.BsonValue
+
+/**
+ * Bson encoder based on the `Json` encoder from the `KSerializer`. To avoid using any
```

Comment:
I think this is a good first step, but would it be possible to support custom types if we just documented that the support is experimental?